core/src/main/java/jenkins/util/io/CompositeIOException.java

```diff
@@ -30,19 +30,52 @@
 import edu.umd.cs.findbugs.annotations.NonNull;
 import java.io.IOException;
 import java.io.UncheckedIOException;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
 @Restricted(NoExternalUse.class)
 public class CompositeIOException extends IOException {
+    private static final long serialVersionUID = 121943141387608148L;
+
+    /**
+     * The maximum number of exceptions that can be reported by a single
+     * {@code CompositeIOException}.
+     * <p>
+     * The number of exceptions is limited to avoid pathological cases where
+     * where a huge number of exceptions could lead to excessive memory usage.
+     * For example, if the number of exceptions was unlimited, a call to
+     * {@code Util.deleteRecursive} could fail with a
+     * {@code CompositeIOException} that contains an exception for every
+     * single file inside of the directory.
+     */
+    public static final int EXCEPTION_LIMIT = 10;
+
     private final List<IOException> exceptions;
 
+    /**
+     * Construct a new {@code CompositeIOException} where the given list of
+     * exceptions are added as suppressed exceptions to the new exception.
+     * <p>
+     * If the given list of exceptions is longer than {@link #EXCEPTION_LIMIT},
+     * the list will be truncated to that length, and an exception whose message
+     * contains the number of exceptions that were removed will be added as a
+     * suppressed exception to the new exception.
+     */
     public CompositeIOException(String message, @NonNull List<IOException> exceptions) {
         super(message);
-        this.exceptions = exceptions;
-        exceptions.forEach(this::addSuppressed);
+        if (exceptions.size() > EXCEPTION_LIMIT) {
+            this.exceptions = new ArrayList<>(exceptions.subList(0, EXCEPTION_LIMIT));
+            this.exceptions.add(new ExceptionLimitReached(exceptions.size() - EXCEPTION_LIMIT));
```

Comment:
I think that is is still important to show users that there were a lot more errors, but IDK if this is the best way to do it. We could also append a message to the message of the `CompositeIOException` itself.

Reply:
better to add to the message imo, seems weird to have another object for this

Reply:
Yeah, my thought was just that it seems kind of random to have the outer error message include the number of exceptions that were removed, and this way you only see that there are unreported errors in contexts were you are looking at a full stack trace, which seems like the only time you would care, but maybe that doesn't really matter.

Any thoughts on the phrasing of the message? Maybe something like "... ($EXCEPTION_LIMIT of N suppressed exceptions were recorded)"?

Reply:
what about?

> Suppressed <#> exceptions

Reply:
Would that number be the total number of exceptions passed to `CompositeIOException` (which are all recorded as suppressed exceptions), or the number that were removed?

Reply:
I would show the # of exceptions that the stacktrace has been discarded for total - exception limit if > 0

Reply:
How about using "discarded" to avoid ambiguity with what "suppressed" refers to?

> Discarded (TOTAL - EXCEPTION_LIMIT) additional exceptions.

Reply:
👍

Reply:
Seems reasonable!

Reply:
I changed to modifying the outer exception's message in 341ffa9.